core/src/ledger/ibc/context/common.rs

```diff
@@ -31,149 +33,426 @@ use crate::ibc::mock::consensus_state::MockConsensusState;
+        self.decode_consensus_state(any)
+    }
+
+    /// Get the client update time
+    fn client_update_time(&self, client_id: &ClientId) -> Result<Timestamp> {
+        let key = storage::client_update_timestamp_key(client_id);
+        match self.read_bytes(&key)? {
+            Some(value) => {
+                let time = TmTime::decode_vec(&value).map_err(|_| {
+                    ContextError::from(ClientError::Other {
                         description: format!(
-                            "Decoding the connection end failed: ID {}",
-                            connection_id,
+                            "Decoding the client update time failed: ID \
+                             {client_id}",
                         ),
                     })
-                })
+                })?;
+                Ok(time.into())
+            }
+            None => Err(ClientError::ClientSpecific {
+                description: format!(
+                    "The client update time doesn't exist: ID {client_id}",
+                ),
             }
-            Ok(None) => Err(ContextError::ConnectionError(
-                ConnectionError::ConnectionNotFound {
-                    connection_id: connection_id.clone(),
-                },
-            )),
-            Err(_) => {
-                Err(ContextError::ConnectionError(ConnectionError::Other {
+            .into()),
+        }
+    }
+
+    /// Store the client update time
+    fn store_update_time(
+        &mut self,
+        client_id: &ClientId,
+        timestamp: Timestamp,
+    ) -> Result<()> {
+        let key = storage::client_update_timestamp_key(client_id);
+        match timestamp.into_tm_time() {
+            Some(time) => self
+                .write_bytes(
+                    &key,
+                    time.encode_vec().expect("encoding shouldn't fail"),
+                )
+                .map_err(ContextError::from),
+            None => Err(ContextError::ClientError(ClientError::Other {
+                description: format!(
+                    "The client timestamp is invalid: ID {client_id}",
+                ),
+            })),
+        }
+    }
+
+    /// Get the client update height
+    fn client_update_height(&self, client_id: &ClientId) -> Result<Height> {
+        let key = storage::client_update_height_key(client_id);
+        match self.read_bytes(&key)? {
+            Some(value) => Height::decode_vec(&value).map_err(|_| {
+                ClientError::Other {
+                    description: format!(
+                        "Decoding the client update height failed: ID \
+                         {client_id}",
+                    ),
+                }
+                .into()
+            }),
+            None => Err(ClientError::ClientSpecific {
+                description: format!(
+                    "The client update height doesn't exist: ID {client_id}",
+                ),
+            }
+            .into()),
+        }
```

Comment:
here are my two cents on this kind of error handling: using matches to handle optionals, if not strictly necessary (e.g. because you need to `return` in one of the branches), introduces unnecessary strain when reading the code. IMO, this is because we are forced to evaluate two possible scenarios mentally (what happens when the code fails? what happens when the code succeeds?). handling errors with methods like `ok_or`, `map_err`, etc, improves readability I think, because errors are handled linearly, if that makes sense, so the logic flows better (e.g. you are not interrupted with error handling).

anyway, this is just a nit

Reply:
Thanks! I modified them with `ok_or`